sled-agent/src/services.rs

```diff
@@ -1305,13 +1301,39 @@ impl ServiceManager {
             // can be supplied - now that we're actively using it, we
             // aren't really handling the "many GZ addresses" case, and it
             // doesn't seem necessary now.
+            info!(self.inner.log, "Zone using its own GZ address as gateway");
             Some(request.zone.gz_addresses[0])
         } else if let Some(info) = self.inner.sled_info.get() {
-            // If the service has not supplied a GZ address, simply add
-            // a route to the sled's underlay address.
-            Some(info.underlay_address)
+            // Only consider a route to the sled's underlay address if the
+            // underlay is up.
+            let sled_underlay_subnet =
+                Ipv6Subnet::<SLED_PREFIX>::new(info.underlay_address);
+
+            if request
+                .zone
+                .addresses
+                .iter()
+                .any(|ip| sled_underlay_subnet.net().contains(*ip))
+            {
+                // If the underlay is up, provide a route to it through an
+                // existing address in the Zone on the same subnet.
+                info!(self.inner.log, "Zone using sled underlay as gateway");
+                Some(info.underlay_address)
+            } else {
+                // If no such address exists in the sled's subnet, don't route
+                // to anything.
+                info!(
+                    self.inner.log,
+                    "Zone not using gateway (even though underlay is up)"
+                );
+                None
+            }
         } else {
             // If the underlay doesn't exist, no routing occurs.
+            info!(
```

Comment:
Will the switch zone eventually get an underlay address and add a route?

Reply:
Good question, and the answer is: Yes, it will.

When we first boot up, the `BootstrapAgent` is responsible for "hardware monitoring", which starts the switch zone iff the tofino is detected:

https://github.com/oxidecomputer/omicron/blob/3e020c907a400e2005874d57fd9e94d8f369e761/sled-agent/src/bootstrap/agent.rs#L397-L404

This calls `HardwareMonitor::new`, which itself has a reference to the `ServiceManager` structure: https://github.com/oxidecomputer/omicron/blob/92258b4d88b292effaf26f361dddb0580b96c1bf/sled-agent/src/bootstrap/hardware.rs#L211 

(Note, this is the **same** `ServiceManager` structure that the `SledAgent` will later use to manage services when we've fully booted)

The bootstrap's `HardwareManager` spawns a tokio task which monitors hardware for the Tofino, and in response, launches the switch zone:

https://github.com/oxidecomputer/omicron/blob/92258b4d88b292effaf26f361dddb0580b96c1bf/sled-agent/src/bootstrap/hardware.rs#L72-L79

**It's important to note -- the `let switch_zone_ip = None` line is critical. If the bootstrap agent is launching the switch zone, it might be doing so before the underlay is up.**

Later on, when the `BootstrapAgent` tries to launch the `SledAgent`. It stops monitoring hardware at this point.

https://github.com/oxidecomputer/omicron/blob/92258b4d88b292effaf26f361dddb0580b96c1bf/sled-agent/src/bootstrap/agent.rs#L526-L527

And when the `SledAgent` starts, it does hardware monitoring for itself:

https://github.com/oxidecomputer/omicron/blob/92258b4d88b292effaf26f361dddb0580b96c1bf/sled-agent/src/sled_agent.rs#L336-L340

If the switch zone already launched, the `SledAgent` will call `activate_switch` with the underlay IP address:

https://github.com/oxidecomputer/omicron/blob/92258b4d88b292effaf26f361dddb0580b96c1bf/sled-agent/src/sled_agent.rs#L353-L365

(this is poking at the same `ServiceManager` used by the bootstrap agent earlier!)

This case is handled explicitly: If the switch zone is running, and needs to be refreshed to add an underlay address and route. It's the same path we use during cold boot / RSS!

https://github.com/oxidecomputer/omicron/blob/2d46f8cc0681c7c25bde57f0fc48ba592d7cd385/sled-agent/src/services.rs#L2742-L2748

Reply:
Aha! It was those last two steps I was missing. Thanks for pointing this out!